API Enforce default_cast

tractorcow opened the pull request:
Replaces #3924

Includes lots of cleanup and additional testing.

The main commit addresses several changes all in one go, which were a bit hard to split into individual commits, as they depend on one another:

```
API Enforce default_cast for all field usages
API Introduce HTMLFragment as casting helper for HTMLText with shortcodes disabled
API Introduce DBField::CDATA for XML file value encoding
API RSSFeed now casts from the underlying model rather than by override
API Introduce CustomMethods::getExtraMethodConfig() to allow metadata to be queried
BUG Remove _call hack from VirtualPage
API Remove FormField::$dontEscape
API Introduce HTMLReadonlyField for non-editable readonly HTML
API FormField::Field() now returns string in many cases rather than DBField instance.
API Remove redundant *_val methods from ViewableData
API ViewableData::obj() no longer has a $forceReturnObject parameter as it always returns an object
BUG  Fix issue with ViewableData caching incorrect field values after being modified.
```

CMS pull request https://github.com/silverstripe/silverstripe-cms/pull/1520

tractorcow commented:
I'd like someone to please do a CMS test; We need to test for UI regressions as a result of this change.

tractorcow commented:
@dhensby noted that we could probably have cast some of the form field properties to Text  instead of HTMLFragment. I'd accept this as a feedback for possible improvement. :D

dhensby commented:
I've run a complete test suite of a default install with this (ie: cms, framework, reports and siteconfig) and they all passed.

dhensby commented:
[holder comment I'll add issues to as I find them]

tractorcow commented:
Probably need to look at fixing shortcodes at the same time:

- https://github.com/silverstripe/silverstripe-framework/issues/2929
- https://github.com/silverstripe/silverstripe-framework/issues/1381
- https://github.com/silverstripe/silverstripe-framework/issues/3925
- https://github.com/silverstripe/silverstripe-framework/issues/5547

Also will fix the regressions you've noted.

dhensby commented:
The cms issue is blocking me testing any further

tractorcow commented:
I've added a new commit for rewriting a bunch of the DBString field summary methods. I haven't got up to implementing these methods for the DBHTMLText field, but I suspect we might simply force these to strip HTML for simplicity. See the @todo

tractorcow commented:
I think I'm happy with the DBText summary methods now... I'm done with the rewrite, so now I'm just testing and fixing.

@dhensby do you want to check I'm going in the right direction?

Rather than attempting to do XML manipulation in an unsafe way, I'm converting HTML fields to plain text for any summary method.

I'm also looking at a proper XML safe implementation based on http://www.pjgalbraith.com/truncating-text-html-with-php/.

tractorcow commented:
Rebased and fixed a few unit tests. Still more tests to resolve:


- [ ] RSSFeedTest
- [ ] DBFieldTest
- [ ] DBHTMLTextTest
- [ ] DBTextTest
- [ ] SSViewerTest
- [ ] ViewableDataTest

tractorcow commented:
Ok, pushed up a fix that fixes ALL of the shortcode issues, except for #5678, which I'll do tomorrow.

tractorcow commented:
Another squash and rebase, cleanup, fix of last minute DB crashing when regenerating Enum columns.... ugh.

tractorcow commented:
TestSession module needs update... it still refers to a removed/deprecated method. ;P

tractorcow commented:
Ok, only broken tests rely on the CMS pr.

@dhensby want to re-review / test to see if the CMS is working now?

sminnee commented on model/FieldType/DBComposite.php:
```diff
@@ -223,11 +223,14 @@ public function hasField($field) {
+	 * @return $this
 	 */
 	public function setField($field, $value, $markChanged = true) {
+		$this->objCacheClear();
```
Do have to clear all cache values here, and not simply the one for key `$field`?

tractorcow replied:
Copy pasted from below:

The problem is we don't know what caches will be invalidated. E.g. if you modify URLSegment, then the Link, RelativeLink, etc caches are all invalid.

The use case is VERY limited, as rarely are objects modified during rendering.

sminnee commented on forms/FormField.php:
```diff
@@ -255,6 +251,22 @@ class FormField extends RequestHandler {
 
+	private static $casting = array(
+		'FieldHolder' => 'HTMLFragment',
+		'Field' => 'HTMLFragment',
```
If Field is already using `renderTemplate()`, which returns an HTMLFragment, why does it casting? Is it just here for the benefit of subclasses that don't use a template?

tractorcow replied:
Because it doesn't always use renderTemplate(). This is a check to ensure it is forced to HTMLFragment in case subclasses don't.

sminnee commented on view/ViewableData.php:
```diff
@@ -384,17 +337,32 @@ protected function objCacheName($fieldName, $arguments) {
+	 *
+	 * @return $this
+	 */
+	protected function objCacheClear() {
```
Are we able to invalidate caches key-by-key? I'd be a bit worried about the performance impacts of clean-sweep caching...

tractorcow replied:
The problem is we don't know what caches will be invalidated. E.g. if you modify URLSegment, then the Link, RelativeLink, etc caches are all invalid.

The use case is VERY limited, as rarely are objects modified during rendering.

sminnee commented on forms/DropdownField.php:
```diff
@@ -110,7 +110,7 @@ protected function getFieldOption($value, $title) {
 	/**
 	 * @param array $properties
-	 * @return HTMLText
+	 * @return string
```
does this return a string?

tractorcow replied:
I'm going to change this to string|HTMLText since there are so many variaties of formfields that either do raw string manipulation or use renderTemplate().

sminnee commented on forms/FormAction.php:
```diff
@@ -83,7 +91,7 @@ public function setFullAction($fullAction) {
 
 	/**
 	 * @param array $properties
-	 * @return HTMLText
```
Doesn't this return HTMLFragment?

sminnee commented on forms/FormAction.php:
```diff
@@ -100,7 +108,7 @@ public function Field($properties = array()) {
 	/**
 	 * @param array $properties
-	 * @return HTMLText
+	 * @return string
```
Doesn't this return HTMLFragment?

sminnee commented on forms/FormField.php:
```diff
@@ -925,7 +939,6 @@ public function setSmallFieldHolderTemplate($smallFieldHolderTemplate) {
 	 * such as an input tag.
 	 *
 	 * @param array $properties
-	 *
```
Doesn't this return HTMLFragment?

tractorcow commented:
Also, The problem is we don't know what caches will be invalidated. E.g. if you modify URLSegment, then the Link, RelativeLink, etc caches are all invalid.

Also, now that https://github.com/silverstripe/silverstripe-framework/pull/4162#event-706478991 has been merged, we can go back to associative arrays for the DBHTMLField options. @dhensby if you merge up I can rebase this (with the fixes @sminnee suggested above) and then we can clean up a few of these casting helpers.

dhensby commented:
I've had a play - looking good. The login form message issue still exists, though

dhensby commented:
Does the `HTMLText::exists` method need updating?

tractorcow commented:
I've done a bunch of refactoring and implemented the feedback I've been given.

I still need to fix some more test regressions, and update the DBString::parseConstructorOptions() to use proper associative arrays, since object::parse_class_spec now properly supports it. :)

tractorcow commented:
I got the new DBHTMLText constructor format working, so now you can use normal PHP array syntax in your db casting helpers.

However, I had too much trouble making Field() (and other) methods to return plain strings instead of DBHTMLText instances, so I've left this behaviour in place.

Once we get the new casting behaviour in, changing these methods can be another task, but I'd rather get this in for now, and not try to do too much at once.

tractorcow commented:
Fixes https://github.com/silverstripe/silverstripe-framework/pull/5798 too

tractorcow commented:
@dhensby fixed the login form problem. Not in the best way I could think of... we still need a way of setting custom castings on customised templates.

Please let's merge this. :)

dhensby commented:
Ok. Codeclimate failures are non-issues

Travis failures are cms problems that will be resolved with the cms pr

dhensby commented:
Top job!

tractorcow commented:
After more than a year of work we are done. :P

dhensby commented:
"done"

tractorcow commented:
Fixes for regressions at https://github.com/silverstripe/silverstripe-cms/pull/1546 and https://github.com/silverstripe/silverstripe-framework/pull/5806
